crates/runtime/src/memory.rs

```diff
@@ -177,10 +164,120 @@ impl RuntimeLinearMemory for MmapMemory {
 
     /// Return a `VMMemoryDefinition` for exposing the memory to compiled wasm code.
     fn vmmemory(&self) -> VMMemoryDefinition {
-        let mut mmap = self.mmap.borrow_mut();
+        let mmap = self.mmap.borrow();
         VMMemoryDefinition {
             base: mmap.alloc.as_mut_ptr(),
             current_length: mmap.size as usize * WASM_PAGE_SIZE as usize,
         }
     }
 }
+
+enum MemoryStorage {
+    Static {
+        base: *mut u8,
+        size: Cell<u32>,
+        maximum: u32,
+        make_accessible: unsafe fn(*mut u8, usize) -> bool,
+    },
+    Dynamic(Box<dyn RuntimeLinearMemory>),
```

Comment:
I was looking to avoid allocating a trait object for the static case, but I'm not opposed to putting this behind the trait to simplify the code here.

Perhaps I was a little overzealous on trying to get the pooling allocator to heap allocate as little as possible to service a module instantiation request.

Reply:
I'd be ok either way myself, I figured it was slightly odd that we have a trait but then we also have an enum to do roughly the same thing.

FWIW I was curious if we could make a test where we use a custom `#[global_allocator]` to do something like count the number of allocations that happens during `Instance::new` while using a pooling allocator, and we could have the test clamp with something like "we perform 10 allocations today and that number must not go up" so we can catch additional allocations (and even update the test as we remove allocations in theory). Felt like a little-bit overkill though too.

Reply:
I'm leaning towards just leaving this as-is and not boxing a trait-object in the static case; I'm of the opinion that the `MemoryStorage` enum doesn't complicate it all *that* much (it's a pretty small amount of code) and not having to box a trait object will be desirable for Lucet-like use cases.  Marking this as resolved.